operator/validator/controller.go

```diff
@@ -350,57 +347,69 @@ func (c *controller) handleRouterMessages() {
 	}
 }
 
-var nonCommitteeValidatorTTLs = map[spectypes.RunnerRole]int{
-	spectypes.RoleCommittee:  64,
-	spectypes.RoleProposer:   4,
-	spectypes.RoleAggregator: 4,
-	//spectypes.BNRoleSyncCommittee:             4,
+var committeeObserverValidatorTTLs = map[spectypes.RunnerRole]int{
+	spectypes.RoleCommittee:                 64,
+	spectypes.RoleProposer:                  4,
+	spectypes.RoleAggregator:                4,
 	spectypes.RoleSyncCommitteeContribution: 4,
 }
 
 func (c *controller) handleWorkerMessages(msg network.DecodedSSVMessage) error {
-	var ncv *committeeObserver
 	ssvMsg := msg.(*queue.SSVMessage)
+	observer := c.getCommitteeObserver(ssvMsg.GetID())
 
-	item := c.getNonCommitteeValidators(ssvMsg.GetID())
-	if item == nil {
-		committeeObserverOptions := validator.CommitteeObserverOptions{
-			Logger:            c.logger,
-			NetworkConfig:     c.networkConfig,
-			ValidatorStore:    c.validatorStore,
-			Network:           c.validatorOptions.Network,
-			Storage:           c.validatorOptions.Storage,
-			FullNode:          c.validatorOptions.FullNode,
-			Operator:          c.validatorOptions.Operator,
-			OperatorSigner:    c.validatorOptions.OperatorSigner,
-			NewDecidedHandler: c.validatorOptions.NewDecidedHandler,
-			AttesterRoots:     c.attesterRoots,
-			SyncCommRoots:     c.syncCommRoots,
-			DomainCache:       c.domainCache,
-		}
-		ncv = &committeeObserver{
-			CommitteeObserver: validator.NewCommitteeObserver(ssvMsg.GetID(), committeeObserverOptions),
-		}
-		ttlSlots := nonCommitteeValidatorTTLs[ssvMsg.MsgID.GetRoleType()]
-		c.committeesObservers.Set(
-			ssvMsg.GetID(),
-			ncv,
-			time.Duration(ttlSlots)*c.beacon.GetBeaconNetwork().SlotDurationSec(),
-		)
-	} else {
-		ncv = item
-	}
-	if err := c.handleNonCommitteeMessages(ssvMsg, ncv); err != nil {
-		return err
+	if err := c.handleCommitteeObserverMessage(ssvMsg, observer); err != nil {
+		return fmt.Errorf("failed to handle committee observer message: %w", err)
 	}
+
 	return nil
 }
 
-func (c *controller) handleNonCommitteeMessages(msg *queue.SSVMessage, ncv *committeeObserver) error {
+func (c *controller) getCommitteeObserver(msgID spectypes.MessageID) *committeeObserver {
+	c.committeesObserversMutex.RLock()
+
+	// Check if the observer already exists
+	existingObserver, ok := c.committeesObservers.Get(msgID)
+	if ok {
+		c.committeesObserversMutex.RUnlock()
+		return existingObserver
+	}
+
+	c.committeesObserversMutex.RUnlock()
+
 	c.committeesObserversMutex.Lock()
 	defer c.committeesObserversMutex.Unlock()
 
```

Comment:
i think it'a potential race condition because another goroutine might also see non-existance in-between our RUnlock -> Lock and then overwrite us later

if we want maximum optimization, we could re-check existance after Lock and exit if it already exists

however if you don't think it's a big deal for performance then we can just do this:
```suggestion
	c.committeesObserversMutex.Lock()
	defer c.committeesObserversMutex.Unlock()

	// Check if the observer already exists
	existingObserver, ok := c.committeesObservers.Get(msgID)
	if ok {
		return existingObserver
	}
```

Reply:
I guess this optimization is not necessary, I tried it before the last one that worked. I'll revert.